### DISCUSSION TITLE
Custom UI - using WixUIValidatePath

### DISCUSSION THREAD

**Discussion by johnsphere** [wixtoolset/issues#8594] [Questions] [answered]
I saw some similar questions, one wasn't answered and the other suggests that I use an app that doesn't run on my machine.

So here I am...

I would like a UI based on WixUI_InstallDir but with an extra page, called PortNumDlg, that has an editbox for the user to enter a port number into. The page would go between LicenseAgreementDlg and InstallDirDlg.

I read these docs: https://wixtoolset.org/docs/tools/wixext/wixui/#addingremovingdialogs which suggests that the recipe is:
- Copy the WixUI_InstallDir  source to my local machine.
- Author my dialog
- Add my dlg as a `<DialogRef Id="PortNumDlg" />`
- Change the Publishes to insert my dlg into the chain

This is all fine but the existing Publish statements refer to custom actions that, I assume, are in the WixToolkit.UI.wixext that I am using.
I am using that extension/nugget/whatever and my initial WixUI_InstallDir  worked so I assume that things are setup right. I am using HeatWave in Visual Studio 2022.

But... I get a linker error:
```
Installer.Wix\OurUI.wxs(27): error WIX0094: The identifier 'CustomAction:WixUIValidatePath' could not be found. Ensure you have typed the reference correctly and that all the necessary inputs are provided to the linker.
2>Done building project "Installer.Wix.wixproj" -- FAILED.
```

I just copied the original source, verbatim.  This is line 27.
`<Publish Dialog="BrowseDlg" Control="OK" Event="DoAction" Value="WixUIValidatePath" Order="3" />
`
[OurUI.zip](https://github.com/user-attachments/files/15968720/OurUI.zip)

I see that the Publish steps that I copied are using `WixUIValidatePath`.

1) Have I missed some vodoo that lets me use the same ones that WixUI_InstallDir  used?

These are my namespaces at the top of the page: 
```
<Wix xmlns="http://wixtoolset.org/schemas/v4/wxs"
     xmlns:ui="http://wixtoolset.org/schemas/v4/wxs/ui">
```

2) Is it impossible to access the custom actions so I have to write my own that do the same job?

3) is there a third alternative that I haven't thought of?

**Comment by johnsphere** [comment]
This isn't an answer, but I found an issue here: https://github.com/wixtoolset/issues/issues/6998

That suggested that I replace `UIRef with ui:WixUI` 

This gave me more errors: Every "normal" element became unrecognized.

Here is a file snippet:
```
<Wix xmlns="http://wixtoolset.org/schemas/v4/wxs"
     xmlns:ui="http://wixtoolset.org/schemas/v4/wxs/ui">
  <Fragment>
	  <ui:WixUI Id="OurUI">
		  <TextStyle Id="WixUI_Font_Normal" FaceName="Tahoma" Size="8" />
		  <TextStyle Id="WixUI_Font_Bigger" FaceName="Tahoma" Size="12" />
		  <TextStyle Id="WixUI_Font_Title" FaceName="Tahoma" Size="9" Bold="yes" />

		  <Property Id="DefaultUIFont" Value="WixUI_Font_Normal" />
		  <Property Id="WixUI_Mode" Value="InstallDir" />

		  <DialogRef Id="BrowseDlg" />
```

Every TextStyle, Property, and DialogRef give me errors.

For example:
```
error WIX0200: The WixUI element contains an unhandled extension element 'TextStyle'. Please ensure that the extension for elements in the 'http://wixtoolset.org/schemas/v4/wxs' namespace has been provided.
```
Isn't that the default namespace? `http://wixtoolset.org/schemas/v4/wxs` what do I need to provide?

**Comment by mrpalide** [comment]
Same issue. It's a little annoying that no docs/example available for WiX4 and WiX5, and converter couldn't cover all things.

**Reply by johnsphere** [reply]
It might just be a little bug, easily fixed. But, yes, the docs don't mention custom actions, and they probably should.

At some stage I will go through more of the kata videos, they are quite good but time consuming.

**Reply by wmanning** [reply, marked as answer]
Does these docs help?
https://wixtoolset.org/docs/fourthree/faqs/#referencing-the-standard-wixui-dialog-sets

For Wix4.
